discovery-provider/src/queries/get_premium_track_signatures.py

```diff
@@ -0,0 +1,337 @@
+    # batch ownership of ERC1155 nfts for a given collection.
+    erc1155_collection_track_map = defaultdict(list)
+    contract_address_token_id_map: Dict[str, Set[int]] = defaultdict(set)
+    for track in erc1155_gated_tracks:
+        contract_address = Web3.toChecksumAddress(
+            track.premium_conditions["nft_collection"]["address"]  # type: ignore
+        )
+        erc1155_collection_track_map[contract_address].append(track.track_id)
+        track_token_id_set = set(map(int, track_token_id_map[track.track_id]))
+        contract_address_token_id_map[contract_address] = contract_address_token_id_map[
+            contract_address
+        ].union(track_token_id_set)
+
+    # Check ownership of nfts from erc1155 collections from given contract addresses,
+    # using all user eth wallets, and generate signatures for corresponding tracks.
+    with concurrent.futures.ThreadPoolExecutor() as executor:
+        future_to_contract_address_map = {
+            executor.submit(
+                _does_user_own_erc1155_nft_collection,
+                contract_address,
+                user_eth_wallets,
+                list(contract_address_token_id_map[contract_address]),
+            ): contract_address
+            for contract_address in list(erc1155_collection_track_map.keys())
+        }
+        for future in concurrent.futures.as_completed(future_to_contract_address_map):
+            contract_address = future_to_contract_address_map[future]
+            try:
+                # Generate premium content signatures for tracks whose
+                # nft collection is owned by the user.
+                if future.result():
+                    for track_id in erc1155_collection_track_map[contract_address]:
+                        track_signature_map[track_id] = get_premium_content_signature(
+                            {
+                                "id": track_id,
+                                "type": "track",
+                                "user_wallet": user_wallet,
+                            }
+                        )
+            except Exception as e:
+                logger.error(
+                    f"Could not future result for erc1155 contract_address {contract_address}. Error: {e}"
+                )
+
+    return track_signature_map
+
+
+# todo: this will be implemented later
+def _get_sol_nft_gated_track_signatures(
+    user_wallet: str,
+    sol_associated_wallets: List[str],
+    tracks: List[Track],
+):
+    return {}
+
+
+# Generates a premium content signature for each of the nft-gated tracks.
+# Return a map of premium track id -> premium content signature.
+def get_nft_gated_premium_track_signatures(
+    user_id: int, track_token_id_map: Dict[int, List[str]]
+):
+    db = db_session.get_db_read_replica()
+    with db.scoped_session() as session:
+        user_wallet = _get_user_wallet(user_id, session)
+        associated_wallets = get_associated_user_wallet({"user_id": user_id})
```

Comment:
could we combine this in a single roundtrip to the db?

Reply:
yup! good idea